lgc/tool/lgc/lgc.cpp

```diff
@@ -0,0 +1,259 @@
+                                        cl::value_desc("filename (\"-\" for stdout)"));
+
+// -pal-abi-version: PAL pipeline ABI version to compile for (default is latest known)
+static cl::opt<unsigned> PalAbiVersion("pal-abi-version", cl::init(0xFFFFFFFF),
+                                       cl::desc("PAL pipeline version to compile for (default latest known)"));
+
+// =====================================================================================================================
+// Checks whether the input data is actually a ELF binary
+//
+// @param data : Input data to check
+static bool isElfBinary(StringRef data) {
+  bool isElfBin = false;
+  if (data.size() >= sizeof(ELF::Elf64_Ehdr)) {
+    auto pHeader = reinterpret_cast<const ELF::Elf64_Ehdr *>(data.data());
+    isElfBin = pHeader->checkMagic();
+  }
+  return isElfBin;
+}
+
+// =====================================================================================================================
+// Checks whether the output data is actually ISA assembler text
+//
+// @param data : Input data to check
+static bool isIsaText(StringRef data) {
+  // This is called by the lgc standalone tool to help distinguish between its three output types of ELF binary,
+  // LLVM IR assembler and ISA assembler. Here we use the fact that ISA assembler is the only one that starts
+  // with a tab character.
+  return data.startswith("\t");
+}
+// =====================================================================================================================
+// Main code of LGC standalone tool
+//
+// @param argc : Count of command-line arguments
+// @param argv : Command-line arguments
+int main(int argc, char **argv) {
+  const char *progName = sys::path::filename(argv[0]).data();
+  LLVMContext context;
+  LgcContext::initialize();
+  cl::ParseCommandLineOptions(argc, argv, "lgc: command-line tool for LGC, the LLPC middle-end compiler\n"
+      "\n"
+      "The lgc tool parses one or more modules of LLVM IR assembler from the input\n"
+      "file(s) and compiles each one using the LGC interface, into AMDGPU ELF or\n"
+      "assembly. Generally, each input module would have been derived by compiling\n"
+      "a shader or pipeline with amdllpc, and using the -emit-lgc option to stop\n"
+      "before running LGC.\n"
+      "\n"
+      "The options section below lists all options available, including all LLVM ones.\n"
+      "The most relevant ones for the lgc utility are:\n"
+      "\n"
+      "  -extract=<index>         - Extract a single input module from multi-module\n"
+      "                             input. Index is 1-based\n"
+      "  -o=<filename>            - Output filename ('-' for stdout); defaults to the input\n"
+      "                             filename with the directory removed and the extension\n"
+      "                             replaced to match the output type\n"
+      "  -mcpu=<cpu-name>         - Target a specific GPU e.g. 'gfx900'\n"
```

Comment:
I understand this is to keep consistent with llc. Is it possible to rename it to other more clear names, such as -gfxip, -gpu.

Reply:
I prefer to keep it as -mcpu as that option already exists in the llvm code that is linked in to lgc. As well as being consistent with llvm tools.